src/System.Management.Automation/engine/parser/Parser.cs

```diff
@@ -4972,29 +5005,82 @@ private StatementAst UsingStatementRule(Token usingToken)
                 if (equalsToken.Kind == TokenKind.Equals)
                 {
                     SkipToken();
-
-                    var aliasToken = NextToken();
-                    if (aliasToken.Kind is TokenKind.EndOfInput or TokenKind.NewLine or TokenKind.Semi)
+                    // We don't allow subnamespaces in namespace aliases
+                    if (kind == UsingStatementKind.Namespace && itemAst.Extent.Text.Contains('.'))
                     {
-                        UngetToken(aliasToken);
-                        ReportIncompleteInput(After(equalsToken),
-                            nameof(ParserStrings.MissingNamespaceAlias),
-                            ParserStrings.MissingNamespaceAlias);
-                        return new ErrorStatementAst(ExtentOf(usingToken, equalsToken));
+                        ReportError(itemAst.Extent, nameof(ParserStrings.NamespaceAliasContainsNamespace), ParserStrings.NamespaceAliasContainsNamespace);
+                        return new ErrorStatementAst(ExtentOf(usingToken, itemAst.Extent));
                     }
 
-                    if (aliasToken.Kind == TokenKind.Comma)
+                    Ast aliasAst;
+                    Token aliasToken;
+                    if (kind is UsingStatementKind.Type)
                     {
-                        ReportError(aliasToken.Extent, nameof(ParserStrings.UnexpectedUnaryOperator), ParserStrings.UnexpectedUnaryOperator, aliasToken.Text);
-                        return new ErrorStatementAst(ExtentOf(usingToken, aliasToken));
+                        // prevents bypassing the type whitelist by overriding an allowed type accelerator with a blocked type like:
+                        // using type string = System.Collections.Generic.List[string]; [string]
+                        if (Runspace.DefaultRunspace?.ExecutionContext?.LanguageMode == PSLanguageMode.ConstrainedLanguage)
```

Comment:
I took this check from the ClassDefinitionRule: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Parser.cs#L4202 I can of course change it if you want but I just figured that it would be okay if we are already doing it with the class definition parsing.

Reply:
Yeah I think the existing check is fine because it'll still hit more CLM checks at runtime. I don't think that'll be the case for this though

Reply:
I'm a little unsure of how and where to add the compiler check. My best guess is somewhere around here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs#L2760 and that I should simply throw an error if the language mode is constrained. I just don't know how to check the language mode from there. The Compiler class doesn't seem to have any relevant properties for this. There is a DefinePowerShellTypes method that uses the parser to do this: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/parser/Compiler.cs#L2691 but the comment also says it's a "short term solution" (lol) so it doesn't seem right to add something new that uses that approach.

Reply:
Yeah... it would be tough. We'd need a way for `LoadUsingsImpl` to generate an expression that will be part of the final body of the script being compiled...

I wonder if it would be easier to just explicitly block overriding any existing type accelerators. @PaulHigin any thoughts on the right move here from a security/CLM perspective?

Reply:
I don't think this check is necessary here at the parser.  In CL mode, type conversion and creation is already restricted at runtime, to a small set of known/approved types.  I think it would be worthwhile to add some extra tests to ensure that type conversion/creation using these new type accelerators are allowed and restricted as expected with CL mode.

Reply:
Actually, thinking about this some more, I don't believe we even need new tests.  There is no new functionality added here, just a mapping of type/using aliases to types, and the existing runtime restrictions continue to apply as before.

Reply:
Yeah you can just remove the check here in general. I commented out the code and tried to bypass:

![image](https://user-images.githubusercontent.com/24977523/236890373-ceb513b1-b39a-4ed9-9277-5518578b115a.png)

Thankfully it's only the actual runtime identity that matters, so an alias doesn't actually bypass the check here.